pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java

```diff
@@ -283,6 +283,21 @@ public interface ConsumerBuilder<T> extends Cloneable {
      */
     ConsumerBuilder<T> messageListener(MessageListener<T> messageListener);
 
+    /**
+     * Set the {@link ThreadPoolProvider} to be used for message listeners of <b>current consumer</b>.
+     * <i>(default: use executor from PulsarClient,
+     * {@link org.apache.pulsar.client.impl.PulsarClientImpl#externalExecutorProvider})</i>.
+     *
+     * <p>The listener thread pool is exclusively owned by current consumer
+     * that are using a "listener" model to get messages. For a given internal consumer,
+     * the listener will always be invoked from the same thread, to ensure ordering.
+     *
+     * <p> The user need to shut down the thread pool after closing the consumer to avoid leaks.
+     * @param listenerThreadsProvider the threads provider of the consumer message listener
+     * @return the consumer builder instance
+     */
+    ConsumerBuilder<T> listenerThreadsProvider(ThreadPoolProvider listenerThreadsProvider);
```

Comment:
I believe that this would be sufficient for the use case:
```suggestion
    ConsumerBuilder<T> messageListenerExecutor(java.util.concurrent.Executor executor);
```

Reply:
It's possible that `java.util.concurrent.Executor` isn't optimal. 

Pulsar client executes Key_Shared subscriptions on different threads:
https://github.com/apache/pulsar/blob/fb80007a47deaadb82d0b1b1e4fcd6ca04c05c9c/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java#L1129-L1136

I guess that's the reason why you have added the `ThreadPoolProvider`. To simplify, I guess one possibility is to have an interface with an execute method that provides the `Message` and the `Runnable`.

Reply:
> It's possible that `java.util.concurrent.Executor` isn't optimal.
> 
> Pulsar client executes Key_Shared subscriptions on different threads:
> 
> https://github.com/apache/pulsar/blob/fb80007a47deaadb82d0b1b1e4fcd6ca04c05c9c/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java#L1129-L1136
> 
> I guess that's the reason why you have added the `ThreadPoolProvider`. To simplify, I guess one possibility is to have an interface with an execute method that provides the `Message` and the `Runnable`.

yeah, in the current Pulsar client versions, we needs to ensure that each consumer's message listener is executed in the same thread to avoid lock contention and ensure message order , so it cannot simply use java.util.concurrent.Executor. And, I will simplify the interface later. Thanks for your idea!